lib/std/os.zig

```diff
@@ -455,7 +455,43 @@ pub fn abort() noreturn {
     if (!builtin.link_libc and builtin.os.tag == .linux) {
         raise(SIG.ABRT) catch {};
 
-        // TODO the rest of the implementation of abort() from musl libc here
+        // If there was a SIGABRT handler installed and it returned, or if
+        // SIGABRT was blocked or ignored, take an Async-Signal-safe lock to
+        // prevent sigaction from installing a new SIGABRT handler, uninstall
+        // any handler that may be present, and re-raise the signal to generate
+        // the default action of abnormal termination.
```

Comment:
- What would be difference between this vs. raising `SIG.KILL` below?
- Couldn't one use sigaction without acquiring said global lock?

Reply:
"SIGKILL signal cannot be caught or ignored and the receiving process cannot perform any clean-up upon receiving this signal. SIGABRT signal can be caught, but it cannot be blocked."
https://serverfault.com/a/873697

The problem is that sigaction is not under our control, so the user may do some above mentioned weird stuff by accident or intentional with multiple threads.

Thus:
1. run user-installed signal handler (not sure, if we support this yet)
2. block signals
3. take a lock to ensure the default abort handler is run by exactly 1 thread
4. run abort again, now with uninstalled signal handler (DFL for default)
5. failsafe mode: only enable receiving ABRT (This part I neither understand, because taking the lock twice is UB.)
6. try atomic hardcrash with nullptr dereference (which caused some weird behavior, so I removed it)
7. still running? ask Kernel to KILL it.
8. still running? beg it it to exit(127).

Reply:
@matu3ba 

1. `raise(ABRT)` would run the user-installed signal handler i'd imagine.
2. user-installed signal handler for ABRT returns from raise(), so we block all signals `sigprocmask(BLOCK)`
3. instead of taking a lock, we should grab an atomic bool and have other threads sleep forever (locking may assume it will eventually be unlocked by the caller thread).
4. `sigaction(DFL)` to set the signal handlers to their default actions, unregistering the user-installing ones.
5. instead of `tkill(current, ABRT)`, should we instead `raise(ABRT)` so that when signals are unblocked, abrt occurs for the entire process?
6. unblock signal handlers (`sigaction(UNBLOCK)`) which runs the `raise/tkill(ABRT)` queued from before and triggers its default handler
7. still running? raise(KILL). This should be uncatchable from what you've explained
8. still running? kernel should've killed the process from raise(KILL). This IMO should be `unreachable` rather than `exit()`.

Reply:
@kprotty

1. yes
2. yes
3. The atomic must be global and conditional, if the code is compiled for single
or multithreaded code. something along (for now SeqCst for simplicity)
```zig
        var abort_entered = std.atomic.Atomic(bool).init(false);
        // must never fail, so strong CAS
        while(abort_entered.compareAndSwap(false, true, .SeqCst, .SeqCst)) {}
```
Unfortunately aweful implementations can call abort() within the abort-handler,
which is why the recursive (un)locking dance in glibc exists.
4. yes
5. raise(ABRT) has more overhead due to (BLOCK, gettid(), tkill, UNBLOCK) and we can use tkill directly instead. (we have already blocked all signals)
6. I'll change the comments. In theory could the signal queue run full, if not implemented
   as ringbuffer, but I think this is unfeasible to assume.
7. yes
8. If the process is not running Kernel code, then it should not be able to continue.
   https://unix.stackexchange.com/a/485657

Context
Signal handlers are per process and masks are per thread.
Accordingly, signal delivery (raise and tkill) is per thread.
https://unix.stackexchange.com/a/597387
https://stackoverflow.com/questions/11679568/signal-handling-with-multiple-threads-in-linux

Linux man page on abort()

> The abort() function first unblocks the SIGABRT signal, and then raises that signal for the
calling process (as though raise(3) was called).

Threads can only block their own signal masks, so this is very poorly phrased.
The man page for raise() writes
"In a multithreaded program it is equivalent to pthread_kill(pthread_self(), sig);"
and for posix
"The effect of the raise() function shall be equivalent to calling: pthread_kill(pthread_self(), sig);".
This reads like musl is wrong on not masking in SIG.ABRT to run the default handler
and instead relies on the application code to "do the correct thing".
However, one could also argue that POSIX is wrong, because the application
should not rely on flawed implementation logic.

> This results in the abnormal termination of the process unless the
SIGABRT signal is caught and the signal handler does not return (see longjmp(3)).

Currently, Zig kinda rules this out with `noreturn` as abort function signature.

> If the SIGABRT signal is ignored, or caught by a handler that returns, the  abort()  function
will still terminate the process. It does this by restoring the default disposition for SIGA‐
BRT and then raising the signal for a second time.

This should be now implemented.

Reply:
I did add your suggestions. 
Please reopen, if you find things discussed here you think are missing.